### ISSUE TITLE
[emu-sv] restore differentiability

### ISSUE THREAD

**Issue by sgrava** [pasqal-io/emulators#127]
Refactoring of the `SVBackend_impl` broke differentiability and managed to get around a test.
 
### Goal
Restore differentiability and specialize test.

### Description

Latest refactor
https://github.com/pasqal-io/emulators/blob/32180897e64f205b78d15513b501a93328acdfe7/emu_sv/sv_backend_impl.py#L174-L178

Introduced this logic to use or not the `autograd.Function` derived stepper in the time evolution.
This was done to separate the two workflows:
1. I will need to compute gradients in the future, so let torch do its magic.
2. I don't need gradient, so just compute the step in the normal way.

While separating the two workflows conceptually nice, in both cases torch will build a computational graph if some Hamiltonian parameter is flagged with require grad.
However, in 2, it will use its built methods. For this reason, when torch spots that we are applying `index_add_`, it will complain since its automatic differentiation does not support such operation. It is for this exact reason that we wrote our own backpropagation which goes around this issue.

Overall, probably the point is that `initial_state.requires_grad` is not sufficient to rule out that you'll need some backpropagation. The only way that comes to my mind is to assert that **all** parameters omega, delta, phi, interactions, initial state, do not require grad. But torch probably do this automatically.

### Actions
- [  ] better test
- [  ] remove logic `if initial_state.requires_grad`
- [  ] check again that torch overhead for running forward without gradient requirements is negligible
